### PR TITLE
feat: regulation worker support for parallel processing

### PR DESCRIPTION
**Pull request by atzoum**
# Description

A single regulation worker instance can now process multiple jobs in parallel

## Notion Ticket

[Link](https://www.notion.so/rudderstacks/8aac9087df644365acdf64e28e290153?v=6e06b0a5ade24f0aa5ffe05dc2972e84&p=e9365dded761444b996aa1e79f2cc80e&pm=s)

## Security

- [x] The code changed/added as part of this pull request won't create any security issues with how the software is being used.

### CODE REVIEW

**Review by saurav-malani — CHANGES_REQUESTED** [review]
We missed to handle the case where two workers/goroutines are trying to delete from same batch deleter (s3 bucket) in that case the same meta file (`rudderDeleteTracker.txt`) that we maintain in s3 would be used by both the workers, which will create problem. 
A simple solution would be to rename the status tracker file with goroutine & worker name (horizontally scaling case).

File: regulation-worker/internal/service/looper.go [hunk 1]
```diff
@@ -21,47 +21,54 @@ type Looper struct {
...
-	}
-	retryDelay, err := getenvInt("RETRY_DELAY_IN_SECONDS", 60)
-	if err != nil {
-		return fmt.Errorf("reading value: %s from env: %s", "INTERVAL_IN_MINUTES", err.Error())
-	}
-
-	for {
+	maxLoopSleep := config.GetDuration("MAX_LOOP_SLEEP", 10, time.Minute)
+	retryDelay := config.GetDuration("RETRY_DELAY", 60, time.Second)
+	workers := config.GetInt("NUM_WORKERS", 5)
```
[the remark this answers is not in the document]
Comment (atzoum) [reply regulation-worker/internal/service/looper.go:26]: 1 worker == 1 goroutine

File: regulation-worker/internal/service/looper.go [hunk 2]
```diff
@@ -21,47 +21,54 @@ type Looper struct {
...
 
-		err := l.Svc.JobSvc(ctx)
+	g, ctx := errgroup.WithContext(ctx)
+	b := make(chan *model.Job, 1)
```
[the remark this answers is not in the document]
Comment (atzoum) [reply regulation-worker/internal/service/looper.go:29]: To be frank, I don't even think we really need a buffered channel after all

**Comment by atzoum** [comment]
> We missed to handle the case where two workers/goroutines are trying to delete from same batch deleter (s3 bucket) in that case the same meta file (`rudderDeleteTracker.txt`) that we maintain in s3 would be used by both the workers, which will create problem. A simple solution would be to rename the status tracker file with goroutine & worker name (horizontally scaling case).

Thanks for pointing that out, I will have a look.

@achettyiitr is there any other potential issue that you can find with running parallel deletions for `S3` or `S3_DATALAKE`?

**Comment by achettyiitr** [comment]
> > We missed to handle the case where two workers/goroutines are trying to delete from same batch deleter (s3 bucket) in that case the same meta file (`rudderDeleteTracker.txt`) that we maintain in s3 would be used by both the workers, which will create problem. A simple solution would be to rename the status tracker file with goroutine & worker name (horizontally scaling case).
> 
> Thanks for pointing that out, I will have a look.
> 
> @achettyiitr is there any other potential issue that you can find with running parallel deletions for `S3` or `S3_DATALAKE`?

cc @abhimanyubabbar

**Comment by atzoum** [comment]
> > > We missed to handle the case where two workers/goroutines are trying to delete from same batch deleter (s3 bucket) in that case the same meta file (`rudderDeleteTracker.txt`) that we maintain in s3 would be used by both the workers, which will create problem. A simple solution would be to rename the status tracker file with goroutine & worker name (horizontally scaling case).
> > 
> > 
> > Thanks for pointing that out, I will have a look.
> > @achettyiitr is there any other potential issue that you can find with running parallel deletions for `S3` or `S3_DATALAKE`?
> 
> cc @abhimanyubabbar

Can we do without a `rudderDeleteTracker.txt`? Why is it important?

**Comment by abhimanyubabbar** [comment]
> > > > We missed to handle the case where two workers/goroutines are trying to delete from same batch deleter (s3 bucket) in that case the same meta file (`rudderDeleteTracker.txt`) that we maintain in s3 would be used by both the workers, which will create problem. A simple solution would be to rename the status tracker file with goroutine & worker name (horizontally scaling case).
> > > 
> > > 
> > > Thanks for pointing that out, I will have a look.
> > > @achettyiitr is there any other potential issue that you can find with running parallel deletions for `S3` or `S3_DATALAKE`?
> > 
> > 
> > cc @abhimanyubabbar
> 
> Can we do without a `rudderDeleteTracker.txt`? Why is it important?

The rudderDeleteTracker file captures the `fileId's` that are already cleaned and can be skipped the next time the tracker starts again after temporarily erroring out.

**Comment by saurav-malani** [comment]
Another scenario that we need to take care of:-

Screenshot 2023-01-02 at 4 52 50 PM

As here two worker/go routine subsequently downloaded the same file. Hence, the cleanup done by `W1` would be overwritten by `W2`.

**Pull request closed without merging**